File: BI-TJV/semestral/rest-client/app/src/main/java/cz/wrzecond/restsys/CreateListener.java
package cz.wrzecond.restsys;

import android.app.Dialog;
import android.content.DialogInterface;
import android.widget.Toast;

import com.google.gson.Gson;

import java.util.function.Consumer;

import cz.wrzecond.restsys.task.HttpMethod;
import cz.wrzecond.restsys.task.HttpRequest;
import cz.wrzecond.restsys.task.HttpResponse;
import wrzecond.ICreateDTO;
import wrzecond.IReadDTO;

public abstract class CreateListener<T extends IReadDTO> implements DialogInterface.OnShowListener {

    // === Properties
    protected Activity<T> activity;
    protected Dialog dialog;
    protected T dto;

    // === Constructor
    public CreateListener (Activity<T> activity, Dialog dialog, T dto) {
        this.activity = activity;
        this.dialog = dialog;
        this.dto = dto;
    }

    protected void scheduleRequest (ICreateDTO dto, String path) {
        HttpRequest request = this.dto == null ? new HttpRequest(path, HttpMethod.POST)
                : new HttpRequest(path + this.dto.getId(), HttpMethod.PATCH);
        request.setBody(new Gson().toJson(dto));
        activity.getService().scheduleTask(getResponseConsumer(path), request);
    }
    private Consumer<HttpResponse> getResponseConsumer (String path) {
        return (response) -> {
            // Show message and close
            Toast.makeText(activity,
                response.getStatus().isSuccess() && response.getLocationHeader() != null ?
                    activity.getString(R.string.create_success,
                    response.getLocationHeader().replace(path, "")) :
                    ( response.getStatus().isSuccess() ? activity.getString(R.string.edit_success) :
                    activity.getString(R.string.create_edit_failure, response.getStatus().name())
            ), Toast.LENGTH_SHORT).show();

            // Hide and reload list if success
            dialog.dismiss();
            if (response.getStatus().isSuccess())
                activity.loadList();
        };
    }

}

File: BI-KOT/semestral/rest-client/app/src/main/java/cz/wrzecond/restsys/CreateListener.kt
package cz.wrzecond.restsys

import android.app.Dialog
import android.content.DialogInterface
import android.widget.Toast
import com.google.gson.Gson
import cz.wrzecond.restsys.task.HttpMethod
import cz.wrzecond.restsys.task.HttpRequest
import wrzecond.IReadDTO
import wrzecond.IUpdateDTO

abstract class CreateListener<T : IReadDTO> (protected val activity: Activity<T>,
protected val dialog: Dialog, protected val dto: T?) : DialogInterface.OnShowListener {
    protected fun scheduleRequest (updateDTO: IUpdateDTO, path: String) {
        // Init request
        val request = if (dto == null) HttpRequest(path, HttpMethod.POST, body = Gson().toJson(updateDTO))
        else HttpRequest("$path${dto.id}", HttpMethod.PATCH, body = Gson().toJson(updateDTO))

        // Fire request
        activity.service.scheduleTask(request) { response ->
            // Init message
            val message = when {
                response.status.isSuccess && response.locationHeader != null ->
                    activity.getString(R.string.create_success, response.locationHeader.replace(path, ""))
                response.status.isSuccess -> activity.getString(R.string.edit_success)
                else -> activity.getString(R.string.create_edit_failure, response.status.name)
            }
            // Show message and dismiss
            Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
            dialog.dismiss()

            // Reload on success
            if (response.status.isSuccess)
                activity.loadList()
        }
    }
}